src/h3lib/include/polygon.h

```diff
@@ -24,13 +24,61 @@
 #include "bbox.h"
 #include "geoCoord.h"
 #include "h3api.h"
+#include "linkedGeo.h"
+
+/** @struct IterableGeoLoop
+ *  @brief  Struct supporting polymorphic loop iteration
+ */
+typedef struct {
+    union {
+        const Geofence* geofence;            ///< optional Geofence
+        const LinkedGeoLoop* linkedGeoLoop;  ///< optional LinkedGeoLoop
+    };
+    int type;  ///< flag for type held by struct
+} IterableGeoLoop;
+
+/** Flag for Geofence type in IterableGeoLoop */
+#define TYPE_GEOFENCE 1
+/** Flag for LinkedGeoLoop type in IterableGeoLoop */
+#define TYPE_LINKED_GEO_LOOP 2
+
+/** Macro: Init iteration vars for either loop version */
+#define INIT_ITERATION                   \
+    int loopIndex = -1;                  \
+    LinkedGeoCoord* currentCoord = NULL; \
+    LinkedGeoCoord* nextCoord = NULL;
+
+/** Macro: Increment polymorphic loop iteration, or break if done.
+ *  Uses if/else rather than switch to support break
+ */
+#define ITERATE(loop, vertexA, vertexB)                                        \
+    if (loop->type == TYPE_GEOFENCE) {                                         \
```

Comment:
Ok, I think I'm missing something, because I can't figure out how an X macro would work without putting the algorithm code into the macro. As discussed in the PR description, I specifically chose not to do that because I'd really rather not put the core algo into a macro. Is there another approach to using X macros I'm missing?

Reply:
You don't have to do a huge amount of algo code within the macro. You can even (ab)use an include file to implement it instead. For example, consider adding a file `genericAlgos.h` or something. Note that you may not use include guards here (`#ifndef GENERIC_ALGOS_H...` nor `#pragma once`).

```
#ifndef TYPE
#error "TYPE must be defined before including this header"
#endif

#ifndef ITERATE
#error "ITERATE must be defined before include this header"
#endif

bool loopContainsPoint ## TYPE(void* container) {
    INIT_LOOP;
    while (true) {
        ...
        ITERATE(container, a, b);
        ...
    }
}

// Same thing for bbox algos

#undef TYPE
#undef ITERATE
```

Then you have this in your implementation file (maybe `polygon.c`):

```
#define TYPE LinkedGeo
#define ITERATE(loop, vertexA, vertexB) ...

#include "genericAlgos.h"

#define TYPE Geofence
#define ITERATE(loop, vertexA, vertexB) ...

#include "genericAlgos.h"
```

Reply:
Just a thought, the above is almost exactly what a C++ template would do.

Reply:
Ok, I can try this.

Reply:
Idk when it's worthwhile to use these somewhat convoluted hacks. If you like what you have, maybe refactor `ITERATE` to be a function that returns two coordinates (probably as out params just like the macro).

Reply:
I had that in an earlier version, but copying the coordinates rather than assigning them was significantly slower... it's possible I could avoid copying, but it'll come out just as hacky I think. I like the potential perf benefit of the macro, and if we can get that in an include file it might be worthwhile.

Reply:
You don't need to copy, just use pointers as arguments. Also, use `inline` in the function signature. Let me test it out and get back to you.